pkg/dns/aws/dns_test.go

```diff
@@ -133,3 +133,86 @@ func Test_zoneIDFromResource(t *testing.T) {
+		expectedElbv2ServiceEndpointEndpoint string
+		expectedRoute53ServiceEndpoint       string
+	}{{
+		name: "default service endpoints, us-east-1",
+		config: Config{
+			Region: "us-east-1",
+		},
+		expectedTaggingServiceEndpoint:       "https://tagging.us-east-1.amazonaws.com",
+		expectedElbServiceEndpointEndpoint:   "https://elasticloadbalancing.us-east-1.amazonaws.com",
+		expectedElbv2ServiceEndpointEndpoint: "https://elasticloadbalancing.us-east-1.amazonaws.com",
+		expectedRoute53ServiceEndpoint:       "https://route53.amazonaws.com",
+	}, {
+		name: "custom service endpoints",
+		config: Config{
+			Region: "us-east-1",
+			ServiceEndpoints: []ServiceEndpoint{
+				// Service endpoints with unrecognized names
+				// are ignored.
+				{Name: "bogus", URL: "http://ignored"},
+				// When the same service endpoint is specified
+				// more than once, the last one is used, until
+				// all service endpoints have been set.
+				{Name: "tagging", URL: "http://overridden1"},
+				{Name: "tagging", URL: "http://overridden2"},
+				{Name: "tagging", URL: "http://x"},
+				{Name: "elasticloadbalancing", URL: "http://y"},
+				{Name: "route53", URL: "http://z"},
+				// Once all service endpoints have been set,
+				// any further entries are ignored.
+				{Name: "tagging", URL: "http://ignored"},
+				{Name: "elasticloadbalancing", URL: "http://ignored"},
+				{Name: "route53", URL: "http://ignored"},
+			},
+		},
+		expectedTaggingServiceEndpoint:       "http://x",
+		expectedElbServiceEndpointEndpoint:   "http://y",
+		expectedElbv2ServiceEndpointEndpoint: "https://elasticloadbalancing.us-east-1.amazonaws.com",
+		expectedRoute53ServiceEndpoint:       "http://z",
+	}, {
+		name: "default service endpoints, GovCloud East",
+		config: Config{
+			Region: "us-gov-east-1",
+		},
+		expectedTaggingServiceEndpoint:       "https://tagging.us-gov-west-1.amazonaws.com",
+		expectedElbServiceEndpointEndpoint:   "https://elasticloadbalancing.us-gov-east-1.amazonaws.com",
+		expectedElbv2ServiceEndpointEndpoint: "https://elasticloadbalancing.us-gov-east-1.amazonaws.com",
+		expectedRoute53ServiceEndpoint:       "https://route53.us-gov.amazonaws.com",
+	}, {
+		name: "default service endpoints, China North",
+		config: Config{
+			Region: "cn-north-1",
+		},
+		expectedTaggingServiceEndpoint:       "https://tagging.cn-northwest-1.amazonaws.com.cn",
+		expectedElbServiceEndpointEndpoint:   "https://elasticloadbalancing.cn-north-1.amazonaws.com.cn",
+		expectedElbv2ServiceEndpointEndpoint: "https://elasticloadbalancing.cn-north-1.amazonaws.com.cn",
+		expectedRoute53ServiceEndpoint:       "https://route53.amazonaws.com.cn",
+	}}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			validateServiceEndpointsFn = func(provider *Provider) error {
+				return nil
+			}
+			provider, err := NewProvider(tc.config, "0.0.0-0")
+			if !assert.NoError(t, err) {
+				return
+			}
+			assert.NotNil(t, provider.route53)
```

Comment:
Why only route53 can potentially be `nil`? Do we need to check other clients?

Reply:
Only the tagging client can be nil.  https://github.com/openshift/cluster-ingress-operator/compare/5ba53aefdbcd097c9fd4c58ac556571c627a169c..8afaffbf8ddbe65565bad52eea6267b615eceec2 adds additional assertions.